### PR TITLE
[Added latency calculations to AMQP records.

### PR DESCRIPTION
**Pull request by noman-xg**
This pull request adds latency calculation logic to AMQP and a corresponding field "latency" is added to the AMQP table to reflect the calculated latency.

NOTE: For asynchronous AMQP frames since calculating latency is not required so i've used zero latency for the async frames.

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR adds latency calculation to AMQP protocol.

#### What is the testplan for the PR:

```
bazle test src/striling/source_connectors/socket_tracer/...

```

image

_truncated for readability_

#### Which issue(s) this PR fixes:

Fixes  https://github.com/pixie-io/pixie/issues/651
#### Does this PR introduce a user-facing change?

```release-note
NONE

```

#### Additional documentation:

```docs
NONE
```

### CODE REVIEW

**Comment by noman-xg** [comment]
@chengruizhe Requesting review.

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 1]
```diff
@@ -1346,6 +1367,8 @@ void SocketTraceConnector::AppendMessage(ConnectorContext* ctx, const ConnTracke
 
   r.Append<r.ColIndex("req_msg")>(entry.req.msg);
   r.Append<r.ColIndex("resp_msg")>(entry.resp.msg);
+  r.Append<r.ColIndex("latency")>(
```
Comment (chengruizhe) [inline src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:1370]: Let's also edit the `amqp_data.pxl` pxl script and add latency as a column.

**Review by chengruizhe — COMMENTED** [review]
Overall looks good.
I tested the changes on the px-sock-shop demo app with the fixes, and it looks great!
Screen Shot 2022-12-15 at 5 09 45 PM

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 2]
```diff
@@ -1086,6 +1086,27 @@ int64_t CalculateLatency(int64_t req_timestamp_ns, int64_t resp_timestamp_ns) {
...
+    return latency_ns;
+  } 
+  
+  if (!resp_synchronous){
+    return latency_ns;
+  }
+  
+  if (req_timestamp_ns > 0 && resp_timestamp_ns > 0) {
+    latency_ns = std::max(resp_timestamp_ns, req_timestamp_ns) - std::min(resp_timestamp_ns, resp_timestamp_ns);
+    LOG_IF_EVERY_N(WARNING, latency_ns < 0, 100)
```
[the remark this answers is not in the document]
Comment (noman-xg) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:1103]: Fixed in 1eefd2fce4466165e63761d8e189de5f6b00ce2d

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc (hunk 1 above)
Comment (noman-xg) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc:1371]: Fixed in 1eefd2fce4466165e63761d8e189de5f6b00ce2d

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (noman-xg) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc]: Fixed in 86b0db24c97d074268342efc03c6399dd80e2262

File: src/stirling/source_connectors/socket_tracer/amqp_table.h [hunk 3]
```diff
@@ -74,6 +74,10 @@ static constexpr DataElement kAMQPElements[] = {
          types::DataType::STRING,
          types::SemanticType::ST_NONE,
          types::PatternType::GENERAL},
+        {"latency", "AMQP latency",
```
[the remark this answers is not in the document]
Comment (noman-xg) [reply src/stirling/source_connectors/socket_tracer/amqp_table.h]: Fixed in 2538f38544b09fb49945bd0f14d9cd9005f0e67c

File: src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc [hunk 4]
```diff
@@ -1086,6 +1086,27 @@ int64_t CalculateLatency(int64_t req_timestamp_ns, int64_t resp_timestamp_ns) {
...
+  if (!req_synchronous){
+    return latency_ns;
+  } 
+  
+  if (!resp_synchronous){
+    return latency_ns;
+  }
+  
+  if (req_timestamp_ns > 0 && resp_timestamp_ns > 0) {
+    latency_ns = std::max(resp_timestamp_ns, req_timestamp_ns) - std::min(resp_timestamp_ns, resp_timestamp_ns);
```
[the remark this answers is not in the document]
Comment (noman-xg) [reply src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc]: Fixed in 2538f38544b09fb49945bd0f14d9cd9005f0e67c

File: src/stirling/source_connectors/socket_tracer/amqp_table.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (noman-xg) [reply src/stirling/source_connectors/socket_tracer/amqp_table.h]: Fixed in 1eefd2fce4466165e63761d8e189de5f6b00ce2d

**Comment by etep** [comment]
@noman-xg ,

We would like to voice a slight preference toward not using mutable variables. Would like to suggest the following small re-write toward that end:
```
int64_t AMQPCalculateLatency(const int64_t req_timestamp_ns, const int64_t resp_timestamp_ns,
                             const bool req_synchronous, const bool resp_synchronous) {
  if(!req_synchronous || !resp_syncrhonous) {
    return 0;
  }

  if(req_timestamp_ns <= 0 || resp_timestamp_ns <= 0) {
    return 0;
  }

  latency_ns = std::max(req_timestamp_ns, resp_timestamp_ns) - std::min(req_timestamp_ns, resp_timestamp_ns);
  return latency_ns;
}
```

NB args flagged as const also.

This surfaces the return value clearly, in the code, for cases where we cannot compute the latency.

**Comment by noman-xg** [comment]
@etep 
Thanks for taking a pass. I've pushed the suggested re-write in 86fae33bd8fd0591ddbb7327ae3e43404861386f

**Comment by oazizi000** [comment]
Thanks @noman-xg! Will pull this in soon.

**Comment by oazizi000** [comment]
Merged in 4ac47cc7a9051b5a5cb7e60b3dc3160720e15085

**Pull request closed without merging**